src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,100 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int EMPTY_SIZE = 0;
+    private static final double CAPACITY_DIVISOR = 1.5;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        grow();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
 
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add to index " + index);
+        }
+        grow();
+        System.arraycopy(elements, index, elements, index + 1, size++ - index);
+        elements[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
 
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexBounds(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexBounds(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexBounds(index);
+        final T element = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size-- - index - 1);
+        elements[size] = null;
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == elements[i]
+                    || (element != null && element.equals(elements[i]))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Not found element in the list");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == EMPTY_SIZE;
+    }
+
+    private void grow() {
+        if (size == elements.length) {
+            int oldCapacity = elements.length;
+            int newCapacity = (int) (oldCapacity + (oldCapacity / CAPACITY_DIVISOR));
+            if (newCapacity < 0) {
+                newCapacity = Integer.MAX_VALUE;
+            }
+            elements = Arrays.copyOf(elements, newCapacity);
+        }
+    }
+
+    private void checkIndexBounds(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't get from index " + index);
```

Comment:
It is not a suitable message, cause you use this check not only in `get` method. I'd recommend you rewrite the message on `throw new ArrayListIndexOutOfBoundsException("Index  " + index + " does not exist");` or something like this.

Reply:
not fixed message for the exception. `"Can't get from index " + index` isn't suitable, cause you use the method `checkIndexBounds` not only in `get` method.